spring-pulsar/src/main/java/org/springframework/pulsar/config/reactive/AbstractReactivePulsarListenerEndpoint.java

```diff
@@ -0,0 +1,234 @@
+import org.springframework.pulsar.listener.adapter.PulsarMessagingMessageListenerAdapter;
+import org.springframework.pulsar.listener.reactive.ReactivePulsarMessageHandler;
+import org.springframework.pulsar.listener.reactive.ReactivePulsarMessageListenerContainer;
+import org.springframework.pulsar.support.MessageConverter;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+
+/**
+ * Base implementation for {@link ReactivePulsarListenerEndpoint}.
+ *
+ * @param <T> Message payload type.
+ * @author Christophe Bornet
+ */
+public abstract class AbstractReactivePulsarListenerEndpoint<T>
+		implements ReactivePulsarListenerEndpoint<T>, BeanFactoryAware, InitializingBean {
+
+	private final LogAccessor logger = new LogAccessor(LogFactory.getLog(getClass()));
+
+	private String subscriptionName;
+
+	private SubscriptionType subscriptionType;
+
+	private SchemaType schemaType;
+
+	private String id;
+
+	private final Collection<String> topics = new ArrayList<>();
+
+	private String topicPattern;
+
+	private BeanFactory beanFactory;
+
+	private BeanExpressionResolver resolver;
+
+	private BeanExpressionContext expressionContext;
+
+	private BeanResolver beanResolver;
+
+	private Boolean autoStartup;
+
+	private Boolean fluxListener;
+
+	private Integer concurrency;
+
+	@Override
+	public void setBeanFactory(BeanFactory beanFactory) throws BeansException {
+		this.beanFactory = beanFactory;
+		if (beanFactory instanceof ConfigurableListableBeanFactory) {
+			this.resolver = ((ConfigurableListableBeanFactory) beanFactory).getBeanExpressionResolver();
+			this.expressionContext = new BeanExpressionContext((ConfigurableListableBeanFactory) beanFactory, null);
+		}
+		this.beanResolver = new BeanFactoryResolver(beanFactory);
+	}
+
+	@Nullable
+	protected BeanFactory getBeanFactory() {
+		return this.beanFactory;
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+		boolean topicsEmpty = getTopics().isEmpty();
+		if (!topicsEmpty && !StringUtils.hasText(getTopicPattern())) {
+			throw new IllegalStateException("Topics or topicPattern must be provided but not both for " + this);
+		}
+	}
+
+	@Nullable
+	protected BeanExpressionResolver getResolver() {
+		return this.resolver;
+	}
+
+	@Nullable
+	protected BeanExpressionContext getBeanExpressionContext() {
+		return this.expressionContext;
+	}
+
+	@Nullable
+	protected BeanResolver getBeanResolver() {
+		return this.beanResolver;
+	}
+
+	public void setSubscriptionName(String subscriptionName) {
+
+		this.subscriptionName = subscriptionName;
+	}
+
+	@Nullable
+	@Override
+	public String getSubscriptionName() {
+		return this.subscriptionName;
+	}
+
+	public void setId(String id) {
+		this.id = id;
+	}
+
+	@Override
+	public String getId() {
+		return this.id;
+	}
+
+	public void setTopics(String... topics) {
+		Assert.notNull(topics, "'topics' must not be null");
+		this.topics.clear();
+		this.topics.addAll(Arrays.asList(topics));
```

Comment:
I doubt this will be done frequent enough to matter, but another option would be to just create a new immutable list on each invocation and replace the other one. I think this is fine as is though.

Reply:
Done